test/rekt/features/apiserversource/data_plane.go

```diff
@@ -0,0 +1,449 @@
+			LabelSelector: &metav1.LabelSelector{MatchLabels: map[string]string{"e2e": "testing"}},
+		}),
+	}
+
+	f.Setup("install ApiServerSource", apiserversource.Install(source, cfg...))
+	f.Setup("ApiServerSource goes ready", apiserversource.IsReady(source))
+
+	examplePodName := feature.MakeRandomK8sName("example")
+
+	// create a pod so that ApiServerSource delivers an event to its sink
+	// event body is similar to this:
+	// {"kind":"Pod","namespace":"test-wmbcixlv","name":"example-axvlzbvc","apiVersion":"v1"}
+	f.Requirement("install example pod",
+		pod.Install(examplePodName,
+			pod.WithImage(exampleImage),
+			pod.WithLabels(map[string]string{"e2e": "testing"})),
+	)
+
+	f.Stable("ApiServerSource as event source").
+		Must("delivers events",
+			eventasssert.OnStore(sink).MatchEvent(
+				test.HasType("dev.knative.apiserver.ref.add"),
+				test.DataContains(`"kind":"Pod"`),
+				test.DataContains(fmt.Sprintf(`"name":"%s"`, examplePodName)),
+			).AtLeast(1))
+
+	return f
+}
+
+// THIS TEST DOES NOT WORK
+//func DoesNotSendEventsForNonLabelMatchingResources() *feature.Feature {
+//	source := feature.MakeRandomK8sName("apiserversource")
+//	sink := feature.MakeRandomK8sName("sink")
+//	f := feature.NewFeatureNamed("Does not send events for label-unmatching resources")
+//
+//	f.Setup("install sink", eventshub.Install(sink, eventshub.StartReceiver))
+//
+//	sacmName := feature.MakeRandomK8sName("apiserversource")
+//	f.Setup("Create Service Account for ApiServerSource with RBAC for v1.Pod resources",
+//		setupAccountAndRoleForPods(sacmName))
+//
+//	cfg := []manifest.CfgFn{
+//		apiserversource.WithServiceAccountName(sacmName),
+//		apiserversource.WithEventMode("Reference"),
+//		apiserversource.WithSink(svc.AsKReference(sink), ""),
+//		apiserversource.WithResources(v1.APIVersionKindSelector{
+//			APIVersion:    "v1",
+//			Kind:          "Pod",
+//			LabelSelector: &metav1.LabelSelector{MatchLabels: map[string]string{"e2e": "something-else"}},
+//		}),
+//	}
+//
+//	f.Setup("install ApiServerSource", apiserversource.Install(source, cfg...))
+//	f.Setup("ApiServerSource goes ready", apiserversource.IsReady(source))
+//
+//	examplePodName := feature.MakeRandomK8sName("example")
+//
+//	// create a pod so that ApiServerSource delivers an event to its sink
+//	// event body is similar to this:
+//	// {"kind":"Pod","namespace":"test-wmbcixlv","name":"example-axvlzbvc","apiVersion":"v1"}
+//	f.Requirement("install example pod",
+//		pod.Install(examplePodName,
+//			pod.WithImage(exampleImage),
+//			pod.WithLabels(map[string]string{"e2e": "testing"})),
+//	)
+//
+//	f.Stable("ApiServerSource as event source").
+//		Must("does not deliver events for unmatched resources", func(ctx context.Context, t feature.T) {
+//			// sleep some time to make sure the sink doesn't actually receive events
+//			// not because reaction time was too short.
+//			time.Sleep(10 * time.Second)
+//			eventasssert.OnStore(sink).MatchEvent(any()).Not()(ctx, t)
+//		})
```

Comment:
This doesn't work. Eventually the event informer created by MagicEnvironment loses its connection with the ApiServer and the test fails.

It works locally though...